File: Trees/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
// LeetCode 105 - Construct Binary Tree from Preorder and Inorder Traversal
// Time Complexity: O(n) | Space Complexity: O(n)
import java.util.HashMap;
import java.util.Map;

public class ConstructBinaryTreeFromPreorderAndInorderTraversal {
    private int[] preorder;
    private final Map<Integer, Integer> inorderIndex = new HashMap<>();

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int length = preorder.length;
        this.preorder = preorder;

        for (int index = 0; index < length; index++) {
            inorderIndex.put(inorder[index], index);
        }

        return dfs(0, 0, length);
    }

    private TreeNode dfs(int preorderStart, int inorderStart, int size) {
        if (size <= 0) {
            return null;
        }

        int rootValue = preorder[preorderStart];
        int splitIndex = inorderIndex.get(rootValue);
        TreeNode left = dfs(preorderStart + 1, inorderStart, splitIndex - inorderStart);
        TreeNode right = dfs(
            preorderStart + 1 + splitIndex - inorderStart,
            splitIndex + 1,
            size - 1 - (splitIndex - inorderStart)
        );

        return new TreeNode(rootValue, left, right);
    }
}

File: Trees/ConstructBinaryTreeFromPreorderAndInorderTraversal.go
// LeetCode 105 - Construct Binary Tree from Preorder and Inorder Traversal
// Time Complexity: O(n) | Space Complexity: O(n)
package main

func buildTree(preorder []int, inorder []int) *TreeNode {
	inorderIndex := make(map[int]int, len(inorder))

	for index, value := range inorder {
		inorderIndex[value] = index
	}

	var dfs func(int, int, int) *TreeNode
	dfs = func(preorderStart int, inorderStart int, size int) *TreeNode {
		if size <= 0 {
			return nil
		}

		rootValue := preorder[preorderStart]
		splitIndex := inorderIndex[rootValue]
		left := dfs(preorderStart+1, inorderStart, splitIndex-inorderStart)
		right := dfs(
			preorderStart+1+splitIndex-inorderStart,
			splitIndex+1,
			size-1-(splitIndex-inorderStart),
		)

		return &TreeNode{Val: rootValue, Left: left, Right: right}
	}

	return dfs(0, 0, len(preorder))
}
